### PR TITLE
Return static 500.30 file on startup fails for in-process

### PR DESCRIPTION
**Pull request by jkotalik**
For #556 .
There was a duplicated check for validating the status of the application in the shim. I removed this as the application should be the one handling setting the http message and status.
This does create a drawback of us needing to allocate an INPROCESS_HANDLER before returning the error page.

### CODE REVIEW

File: src/AspNetCoreModuleV2/InProcessRequestHandler/inprocessapplication.cpp [hunk 1]
```diff
@@ -29,6 +29,7 @@ IN_PROCESS_APPLICATION::IN_PROCESS_APPLICATION(
 
     InitializeSRWLock(&m_srwLock);
     m_pConfig = pConfig;
+    html500Page = GetStaticHtml500Content();
```
Comment (mikeharder) [inline src/AspNetCoreModuleV2/InProcessRequestHandler/inprocessapplication.cpp]: Why store the error HTML in a member, rather than calling `GetStaticHtml500Content()` every time the error HTML is needed?  If it's for perf, this seems unnecessary since the HTML is only used in error cases.

File: src/AspNetCoreModuleV2/AspNetCore/src/proxymodule.cxx [hunk 2]
```diff
@@ -152,16 +152,6 @@ ASPNET_CORE_PROXY_MODULE::OnExecuteRequestHandler(
 
     m_pApplicationInfo->ExtractApplication(&pApplication);
 
-    // make sure application is in running state
```
Comment (pakrym) [inline src/AspNetCoreModuleV2/AspNetCore/src/proxymodule.cxx]: Why was this removed?

File: src/AspNetCoreModuleV2/InProcessRequestHandler/inprocesshandler.cpp [hunk 3]
```diff
@@ -30,18 +30,28 @@ IN_PROCESS_HANDLER::OnExecuteRequestHandler()
 
     if (FAILED(hr))
     {
-        // TODO remove com_error?
-        /*_com_error err(hr);
-        if (ANCMEvents::ANCM_START_APPLICATION_FAIL::IsEnabled(m_pW3Context->GetTraceContext()))
+        if (!m_pApplication->QueryConfig()->QueryDisableStartUpErrorPage())
```
Comment (pakrym) [inline src/AspNetCoreModuleV2/InProcessRequestHandler/inprocesshandler.cpp]: Can we return a different IAPPLICATION instance if app failed to start?

File: src/AspNetCoreModuleV2/AspNetCore/src/proxymodule.cxx (hunk 2 above)
Comment (jkotalik) [reply src/AspNetCoreModuleV2/AspNetCore/src/proxymodule.cxx]: We shouldn't be checking the status of the application in the shim. This caused an IIS 500 to be returned instead of ours.

File: src/AspNetCoreModuleV2/InProcessRequestHandler/inprocesshandler.cpp (hunk 3 above)
Comment (jkotalik) [reply src/AspNetCoreModuleV2/InProcessRequestHandler/inprocesshandler.cpp]: Not without some refactoring. Here is the current order of operations.
- CreateApplication
- CreateHandler
- OnExecuteRequestHandler
- LoadApplication

I think we need to move LoadApplication before creating the handler. I'll take a quick stab at it and see how hard it is.

Comment (jkotalik) [reply src/AspNetCoreModuleV2/InProcessRequestHandler/inprocesshandler.cpp]: Will be fairly difficult. Let's not do this in this PR IMO.

File: src/AspNetCoreModuleV2/InProcessRequestHandler/inprocessapplication.cpp (hunk 1 above)
Comment (mikeharder) [reply src/AspNetCoreModuleV2/InProcessRequestHandler/inprocessapplication.cpp]: I still wouldn't bother caching this but I agree there's not a big difference either way.

**Comment by jkotalik** [comment]
🆙 📅 . Instead of returning the exception in the inprocess application, we create a new application that constantly returns 500.30s.

**Comment by pan-wang** [comment]
Can we fix the flaky CI  build /tests?   A lot of time,  I don't know the Whether  PR had issue or not and should I look at a PR which has CI failure.

**Comment by jkotalik** [comment]
This is my failure. Investigating.

**Comment by pakrym** [comment]
> Can we fix the flaky CI build /tests? A lot of time, I don't know the Whether PR had issue or not and should I look at a PR which has CI failure.

Tests don't seem to flaky, it's just PRs are breaking them.

File: src/AspNetCoreModuleV2/AspNetCore/src/proxymodule.cxx (hunk 2 above)
Comment (pakrym) [reply src/AspNetCoreModuleV2/AspNetCore/src/proxymodule.cxx]: Should we bring this back since we return IAPPLCIATION instance now?

File: src/AspNetCoreModuleV2/InProcessRequestHandler/StartupExceptionApplication.cpp [hunk 4]
```diff
@@ -0,0 +1,20 @@
...
+#include "StartupExceptionApplication.h"
+
+VOID StartupExceptionApplication::ShutDown()
+{
+    return VOID();
+}
+
+VOID StartupExceptionApplication::Recycle()
+{
+    return VOID();
```
Comment (pakrym) [inline src/AspNetCoreModuleV2/InProcessRequestHandler/StartupExceptionApplication.cpp]: Are you sure this should be empty?

File: test/IISIntegration.FunctionalTests/Inprocess/StartupExceptionTests.cs [hunk 5]
```diff
@@ -50,5 +50,20 @@ public class StartupExceptionTests : IISFunctionalTestBase
 
             Assert.Contains(TestSink.Writes, context => context.Message.Contains(new string('a', 4096)));
         }
+
+
+        [Fact]
+        public async Task Gets500_30_ErrorPage()
+        {
+            var deploymentParameters = Helpers.GetBaseDeploymentParameters("StartupExceptionWebsite");
+            deploymentParameters.PublishApplicationBeforeDeployment = true;
```
Comment (pakrym) [inline test/IISIntegration.FunctionalTests/Inprocess/StartupExceptionTests.cs]: why?

**Comment by mikeharder** [comment]
@pakrym, @pan-wang: We have seen some flaky tests in IISIntegration: https://github.com/aspnet/iisintegration/issues?utf8=%E2%9C%93&q=is%3Aissue+is%3Aopen+flaky.  We do need to fix them eventually but they may not currently be the highest priority.

File: src/AspNetCoreModuleV2/InProcessRequestHandler/StartupExceptionApplication.cpp (hunk 4 above)
Comment (jkotalik) [reply src/AspNetCoreModuleV2/InProcessRequestHandler/StartupExceptionApplication.cpp]: Good point. Let me think about it.

Comment (pan-wang) [reply src/AspNetCoreModuleV2/InProcessRequestHandler/StartupExceptionApplication.cpp]: I am sure about the value of this change. We introduce multiple classes just for handling one error page. Why not just a simple UTILITY method?

Comment (jkotalik) [reply src/AspNetCoreModuleV2/InProcessRequestHandler/StartupExceptionApplication.cpp]: If we want to update the error page in the future, we would need to modify the shim. If we want to use a utility function, we will effectively reuse the inprocess application loaded to just return 500s.

**Comment by jkotalik** [comment]
@dotnet-bot test this please

File: src/AspNetCoreModuleV2/InProcessRequestHandler/StartupExceptionApplication.cpp [hunk 6]
```diff
@@ -0,0 +1,31 @@
...
+    {
+        // IIS scenario.
+        // notify IIS first so that new request will be routed to new worker process
+        m_pHttpServer->RecycleProcess(L"AspNetCore InProcess Recycle Process on Demand");
+    }
+    else
+    {
+        // IISExpress scenario
+        // Shutdown the managed application and call exit to terminate current process
+        exit(0);
```
Comment (pakrym) [inline src/AspNetCoreModuleV2/InProcessRequestHandler/StartupExceptionApplication.cpp]: did you copy it from somewhere? We should avoid doing exit(0) and allow IIS to gracefully stop.

Comment (jkotalik) [reply src/AspNetCoreModuleV2/InProcessRequestHandler/StartupExceptionApplication.cpp]: Yeah we shouldn't need it at all.

**Comment by jkotalik** [comment]
Current code path works fine on IISExpress and IIS during recycle.

**Comment by jkotalik** [comment]
I'll check appoffline scenarios too.

**Comment by jkotalik** [comment]
There seems to be larger issues that should be handled in other PRs, including #950 and #949 . This handles all scenarios we expect it to.

File: src/AspNetCoreModuleV2/InProcessRequestHandler/InProcessApplicationBase.h [hunk 7]
```diff
@@ -0,0 +1,24 @@
...
+{
+public:
+
+    InProcessApplicationBase(IHttpServer* pHttpServer);
+
+    ~InProcessApplicationBase() = default;
+
+    __override VOID Recycle(VOID);
+
+    BOOL m_fRecycleCalled;
```
Comment (pakrym) [inline src/AspNetCoreModuleV2/InProcessRequestHandler/InProcessApplicationBase.h]: private

File: src/AspNetCoreModuleV2/InProcessRequestHandler/InProcessApplicationBase.h [hunk 8]
```diff
@@ -0,0 +1,24 @@
...
+
+class InProcessApplicationBase : public APPLICATION
+{
+public:
+
+    InProcessApplicationBase(IHttpServer* pHttpServer);
+
+    ~InProcessApplicationBase() = default;
+
+    __override VOID Recycle(VOID);
```
Comment (pakrym) [inline src/AspNetCoreModuleV2/InProcessRequestHandler/InProcessApplicationBase.h]: override

File: src/AspNetCoreModuleV2/InProcessRequestHandler/dllmain.cxx [hunk 9]
```diff
@@ -104,11 +107,17 @@ CreateApplication(
         {
             goto Finished;
         }
+        disableStartupPage = pConfig->QueryDisableStartUpErrorPage();
 
         pApplication = new IN_PROCESS_APPLICATION(pServer, pConfig);
-
         pConfig = NULL;
 
+        if (FAILED(pApplication->LoadManagedApplication()))
```
Comment (pakrym) [inline src/AspNetCoreModuleV2/InProcessRequestHandler/dllmain.cxx]: This would cause a double free because of  IN_PROCESS_APPLICATION unique_ptr

File: src/AspNetCoreModuleV2/InProcessRequestHandler/dllmain.cxx [hunk 10]
```diff
@@ -90,46 +93,27 @@ CreateApplication(
...
-
-Finished:
-    if (FAILED(hr))
-    {
-        if (pApplication != NULL)
+        if (FAILED(pApplication->LoadManagedApplication()))
         {
-            delete pApplication;
-            pApplication = NULL;
+            *ppApplication = new StartupExceptionApplication(pServer, disableStartupPage);
```
Comment (pakrym) [inline src/AspNetCoreModuleV2/InProcessRequestHandler/dllmain.cxx]:
```
pApplication  = make_unique<StartupExceptionApplication>();
```

and then `*ppApplication = pApplication.release();` after if

Comment (jkotalik) [reply src/AspNetCoreModuleV2/InProcessRequestHandler/dllmain.cxx]: Can't do that exactly as pApplication is of type `std::unique_ptr<InProcessApplication>`. Would you prefer I don't call new?

Comment (jkotalik) [reply src/AspNetCoreModuleV2/InProcessRequestHandler/dllmain.cxx]: Need to call `LoadManagedApplication()` on the InProcessApplication.

File: test/IISIntegration.FunctionalTests/Inprocess/StartupExceptionTests.cs [hunk 11]
```diff
@@ -54,5 +53,19 @@ public class StartupExceptionTests : IISFunctionalTestBase
...
+
+        [Fact]
+        public async Task Gets500_30_ErrorPage()
+        {
+            var deploymentParameters = Helpers.GetBaseDeploymentParameters("StartupExceptionWebsite");
+
+            var deploymentResult = await DeployAsync(deploymentParameters);
+
+            var response = await deploymentResult.HttpClient.GetAsync("/");
+            Assert.False(response.IsSuccessStatusCode);
```
Comment (pakrym) [inline test/IISIntegration.FunctionalTests/Inprocess/StartupExceptionTests.cs]: Assert err code
